### ISSUE TITLE
VIDEOS

### ISSUE THREAD

**Issue by anto2506** [anto2506/IA-VIDEO#1]
import os
import numpy as np
import tensorflow as tf
from tensorflow.keras import layers, models
import cv2
import matplotlib.pyplot as plt

# --- Paramètres ---
video_length = 16  # Nombre de frames par vidéo
frame_height = 64   # Hauteur des frames
frame_width = 64    # Largeur des frames
channels = 3        # 3 pour RGB
batch_size = 32
epochs = 1000
latent_dim = 100    # Dimension de l'espace latent

# --- Générateur (vidéo) ---
def build_generator():
    model = models.Sequential()

    # Couche d'entrée : bruit aléatoire + dimension temporelle
    model.add(layers.InputLayer(input_shape=(latent_dim + video_length,)))

    # Couches denses pour générer une séquence de frames
    model.add(layers.Dense(256, activation='relu'))
    model.add(layers.Reshape((video_length, 256)))
    model.add(layers.LSTM(256, return_sequences=True))
    model.add(layers.BatchNormalization())

    # Générer chaque frame
    model.add(layers.TimeDistributed(layers.Dense(128, activation='relu')))
    model.add(layers.TimeDistributed(layers.Dense(frame_height * frame_width * channels, activation='tanh')))
    model.add(layers.Reshape((video_length, frame_height, frame_width, channels)))

    return model

# --- Discriminateur (vidéo) ---
def build_discriminator():
    model = models.Sequential()

    # Entrée : séquence de frames
    model.add(layers.InputLayer(input_shape=(video_length, frame_height, frame_width, channels)))

    # Couches convolutionnelles 3D pour traiter la dimension temporelle
    model.add(layers.TimeDistributed(layers.Conv2D(64, (3, 3), strides=(2, 2), padding='same')))
    model.add(layers.TimeDistributed(layers.LeakyReLU(alpha=0.2)))
    model.add(layers.TimeDistributed(layers.Conv2D(128, (3, 3), strides=(2, 2), padding='same')))
    model.add(layers.TimeDistributed(layers.LeakyReLU(alpha=0.2)))
    model.add(layers.TimeDistributed(layers.Flatten()))

    # LSTM pour traiter la séquence temporelle
    model.add(layers.LSTM(128))
    model.add(layers.Dense(1, activation='sigmoid'))

    return model

# --- Construire le GAN ---
generator = build_generator()
discriminator = build_discriminator()

# Optimiseurs
optimizer = tf.keras.optimizers.Adam(0.0002, 0.5)

# Compiler le discriminateur
discriminator.compile(loss='binary_crossentropy', optimizer=optimizer, metrics=['accuracy'])

# GAN combiné (générateur + discriminateur gelé)
discriminator.trainable = False
gan_input = layers.Input(shape=(latent_dim + video_length,))
generated_video = generator(gan_input)
validity = discriminator(generated_video)
gan = models.Model(gan_input, validity)
gan.compile(loss='binary_crossentropy', optimizer=optimizer)

# --- Fonction pour générer une vidéo ---
def generate_video(generator, latent_dim, video_length, frame_height, frame_width, channels):
    noise = np.random.normal(0, 1, (1, latent_dim + video_length))
    generated_video = generator.predict(noise, verbose=0)
    generated_video = (generated_video * 0.5 + 0.5) * 255  # Normaliser entre 0-255
    generated_video = np.uint8(generated_video[0])

    # Sauvegarder la vidéo
    output_path = "generated_video.mp4"
    fourcc = cv2.VideoWriter_fourcc(*'mp4v')
    out = cv2.VideoWriter(output_path, fourcc, 5, (frame_width, frame_height))

    for frame in generated_video:
        out.write(frame)
    out.release()
    print(f"Vidéo générée : {output_path}")
    return output_path

# --- Entraînement (simplifié) ---
def train_gan(generator, discriminator, gan, epochs, batch_size):
    # Dans un vrai projet, il faudrait charger un dataset de vidéos
    # Ici, on simule un entraînement avec du bruit
    for epoch in range(epochs):
        # Générer des vidéos faux
        noise = np.random.normal(0, 1, (batch_size, latent_dim + video_length))
        generated_videos = generator.predict(noise, verbose=0)

        # Générer des labels (0 = faux, 1 = vrai)
        valid = np.ones((batch_size, 1))
        fake = np.zeros((batch_size, 1))

        # Entraîner le discriminateur
        d_loss_real = discriminator.train_on_batch(np.random.rand(batch_size, video_length, frame_height, frame_width, channels), valid)
        d_loss_fake = discriminator.train_on_batch(generated_videos, fake)
        d_loss = 0.5 * np.add(d_loss_real, d_loss_fake)

        # Entraîner le générateur
        noise = np.random.normal(0, 1, (batch_size, latent_dim + video_length))
        g_loss = gan.train_on_batch(noise, valid)

        # Afficher les progrès
        if epoch % 100 == 0:
            print(f"Epoch {epoch}, D Loss: {d_loss[0]}, G Loss: {g_loss}")

# --- Exécuter ---
if __name__ == "__main__":
    print("Début de l'entraînement...")
    train_gan(generator, discriminator, gan, epochs, batch_size)
    print("Génération d'une vidéo...")
    generate_video(generator, latent_dim, video_length, frame_height, frame_width, channels)
